### PR TITLE
Copy liveshare client into external access

### PR DESCRIPTION
**Pull request by dibarbet**
Feature branch.
Copy the client side handlers from liveshare to external access.

Note that everything in the external access project is just a straight copy of the liveshare client side implementation, with very minor changes where necessary to get them working inside roslyn.  The client side will be removed as soon as possible (once the content type issue with liveshare and razor is resolved).  These have been working for over a year already.

Also made a couple fixes to goto def and code actions on the server side to fix issues with older liveshare implementations.

TODO - Pull in fix for gotodef, find references once liveshare removes dependencies.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/CustomProtocol/RunCodeActionParams.cs [hunk 1]
```diff
@@ -6,20 +6,15 @@ namespace Microsoft.CodeAnalysis.LanguageServer.CustomProtocol
...
-    /// This is an implementation detail of the server that is passed to the clients
-    /// and returned back without the clients parsing it, so no need to make it public.
+    /// Unfortunately, while the client does not use these params, it gets parsed on the client side.
+    /// Therefore this type must match the client's type.
     /// </summary>
     internal class RunCodeActionParams
     {
         /// <summary>
-        /// The original text document.
+        /// Params that were passed to originally get a list of codeactions.
```
Comment (heejaechang) [inline src/Features/LanguageServer/Protocol/CustomProtocol/RunCodeActionParams.cs]: probably work item to remove this in client side and fix this parameter up later?

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 2]
```diff
@@ -57,6 +57,20 @@ internal static class ProtocolConversions
             { WellKnownTags.NuGet, LSP.CompletionItemKind.Text }
         };
 
+        public static LSP.TextDocumentPositionParams PositionToTextDocumentPositionParams(int position, SourceText text, Document document)
```
Comment (heejaechang) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs]: methods in this file. should we all convert to extension method?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs [hunk 3]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
 using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
```
Comment (heejaechang) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs]: sort using?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs [hunk 4]
```diff
@@ -45,7 +46,14 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
                 // If we have a codeaction with a single applychangesoperation, we want to send the codeaction with the edits.
                 var operations = await codeAction.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
 
-                var clientSupportsWorkspaceEdits = clientCapabilities?.Workspace?.WorkspaceEdit?.DocumentChanges == true;
+                var clientSupportsWorkspaceEdits = true;
```
Comment (heejaechang) [inline src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs]: extension method?

File: src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs [hunk 5]
```diff
@@ -10,14 +10,14 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
...
-    internal class GoToDefinitionHandler : GoToDefinitionHandlerBase, IRequestHandler<LSP.TextDocumentPositionParams, LSP.Location[]>
+    internal class GoToDefinitionHandler : GoToDefinitionHandlerBase, IRequestHandler<LSP.TextDocumentPositionParams, object>
     {
         [ImportingConstructor]
         public GoToDefinitionHandler(IMetadataAsSourceFileService metadataAsSourceFileService) : base(metadataAsSourceFileService)
         {
         }
 
-        public async Task<LSP.Location[]> HandleRequestAsync(Solution solution, LSP.TextDocumentPositionParams request,
+        public async Task<object> HandleRequestAsync(Solution solution, LSP.TextDocumentPositionParams request,
```
Comment (heejaechang) [inline src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs]: work item to track whether we want to leave things as object or start use UnionType people are talking about on LSP

File: src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs [hunk 6]
```diff
@@ -26,13 +27,20 @@ void M()
 }";
             var (solution, locations) = CreateTestSolution(markup);
             var expected = CreateCommand(CSharpFeaturesResources.Use_implicit_type, locations["caret"].First());
+            var clientCapabilities = new LSP.ClientCapabilities()
```
Comment (heejaechang) [inline src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs]: can we have a helper extension method to do this simpler? by the way, this doesnt work? jtoken["Experimental"]["supportsWorkspaceEdits"] = false?

you can get jtoken by JToken.FromObject(new Lsp.ClientCapabilities()) ?

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs (hunk 4 above)
Comment (heejaechang) [reply src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs]: I think you can also do.
clientCapabilitiesExtensions.SelectToken("supportsWorkspaceEdits")?.Value<bool>() ?

File: src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Classification
+{
+    [ExportLanguageServiceFactory(typeof(IClassificationService), StringConstants.CSharpLspLanguageName), Shared]
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs]: do we need one for typescript until they move out?

Comment (heejaechang) [reply src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs]: and work item to move those to typescript external access dll once they create one.

File: src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.Exports.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.CodeActions
+{
+    [Shared]
+    [ExportCodeRefactoringProvider(StringConstants.CSharpLspLanguageName)]
+    internal class CSharpLspCodeActionProvider : RoslynCodeActionProvider
+    {
+        [ImportingConstructor]
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.Exports.cs]: probably typescript never use this?

File: src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        public RoslynCodeActionProvider(RoslynLSPClientServiceFactory roslynLSPClientServiceFactory, IDiagnosticAnalyzerService diagnosticAnalyzerService)
+        {
+            _roslynLSPClientServiceFactory = roslynLSPClientServiceFactory ?? throw new ArgumentNullException(nameof(roslynLSPClientServiceFactory));
+            _diagnosticAnalyzerService = diagnosticAnalyzerService ?? throw new ArgumentNullException(nameof(diagnosticAnalyzerService));
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            // This provider is exported for all workspaces - so limit it to just our workspace.
+            if (context.Document.Project.Solution.Workspace.Kind != WorkspaceKind.AnyCodeRoslynWorkspace)
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.cs]: I guess all these are just straight port from live share implementation?

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.Exports.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Completion
+{
+    [ExportCompletionProvider("CSharpLspCompletionProvider", StringConstants.CSharpLspLanguageName), Shared]
+    internal class CSharpLspCompletionProvider : RoslynCompletionProvider
+    {
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.Exports.cs]: same typescript.

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.cs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        public RoslynCompletionProvider(RoslynLSPClientServiceFactory roslynLSPClientServiceFactory)
+        {
+            _roslynLSPClientServiceFactory = roslynLSPClientServiceFactory ?? throw new ArgumentNullException(nameof(roslynLSPClientServiceFactory));
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            // This provider is exported for all workspaces - so limit it to just our workspace & the debugger's intellisense workspace
+            if (context.Document.Project.Solution.Workspace.Kind != WorkspaceKind.AnyCodeRoslynWorkspace &&
+                context.Document.Project.Solution.Workspace.Kind != StringConstants.DebuggerIntellisenseWorkspaceKind)
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.cs]: seems very arbitrary. why some just checked AnyCode and some checks debugger as well. but, these will be removed so, probably okay.

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.Exports.cs (hunk 10 above)
Comment (heejaechang) [reply src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.Exports.cs]: ah. just remembered that typescript actually use VSSDK so they don't need it. sorry. ignore.

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionService.Exports.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+            return new RoslynCompletionService(languageServices.WorkspaceServices.Workspace,
+                originalService, StringConstants.VBLspLanguageName);
+        }
+    }
+
+    // In d16, this export is to disable local TypeScript completion within a Live Share session; it is not used
+    [ExportLanguageServiceFactory(typeof(CompletionService), StringConstants.TypeScriptLanguageName, WorkspaceKind.AnyCodeRoslynWorkspace), Shared]
+    internal class TypeScriptLspCompletionServiceFactory : ILanguageServiceFactory
+    {
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionService.Exports.cs]: we probably want to have work item to move this to typescript ExternalAccess dll once they create one.

File: src/Tools/ExternalAccess/LiveShare/Projects/RoslynRemoteProjectInfoProvider.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        private const string SystemUriSchemeExternal = "vslsexternal";
+
+        private readonly RoslynLSPClientServiceFactory _roslynLSPClientServiceFactory;
+        //private readonly IVsRemoteWorkspaceManager _remoteWorkspaceManager;
+
+        [ImportingConstructor]
+        public RoslynRemoteProjectInfoProvider(RoslynLSPClientServiceFactory roslynLSPClientServiceFactory)//, IVsRemoteWorkspaceManager remoteWorkspaceManager)
+        {
+            _roslynLSPClientServiceFactory = roslynLSPClientServiceFactory ?? throw new ArgumentNullException(nameof(roslynLSPClientServiceFactory));
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Projects/RoslynRemoteProjectInfoProvider.cs]: so, what is current plan for this? this is not part of LSP. so we need to talk with project system people on this I guess?

File: src/Tools/ExternalAccess/LiveShare/Rename/RoslynRenameService.Exports.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Rename
+{
+    [ExportLanguageService(typeof(IEditorInlineRenameService), StringConstants.CSharpLspLanguageName), Shared]
+    internal class CSharpLspRenameService : RoslynRenameService
```
Comment (heejaechang) [inline src/Tools/ExternalAccess/LiveShare/Rename/RoslynRenameService.Exports.cs]: hmm.. wondering what we should do this for LSP. probably work item needed?

File: src/Features/LanguageServer/Protocol/CustomProtocol/RunCodeActionParams.cs (hunk 1 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/CustomProtocol/RunCodeActionParams.cs]: https://github.com/dotnet/roslyn/projects/45#card-20619668

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs]: in a previous PR @tmat requested they not be extensions.
https://github.com/dotnet/roslyn/pull/34794#discussion_r274064637

File: src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs (hunk 5 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs]: Sure.  But for these they will have to be object to be backwards compatible with older liveshare clients.

File: src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs (hunk 7 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs]: Nope, there's only a couple places where the roslyn client is exporting typescript for clients newer than 16.0

https://github.com/dotnet/roslyn/projects/45#card-21807510

File: src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.Exports.cs (hunk 8 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.Exports.cs]: yep typescript liveshare client is not using this in 16.0

File: src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.cs (hunk 9 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/CodeActions/RoslynCodeActionProvider.cs]: Yep straight copy, with some minor edits.

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.cs (hunk 11 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionProvider.cs]: I'm not sure - I can dig into it but these will all hopefully be deleted soon.

File: src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionService.Exports.cs (hunk 12 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/Completion/RoslynCompletionService.Exports.cs]: https://github.com/dotnet/roslyn/projects/45#card-21807510

File: src/Tools/ExternalAccess/LiveShare/Rename/RoslynRenameService.Exports.cs (hunk 14 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/Rename/RoslynRenameService.Exports.cs]: https://github.com/dotnet/roslyn/projects/45#card-20619668

File: src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs (hunk 5 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Definitions/GoToDefinitionHandler.cs]: https://github.com/dotnet/roslyn/projects/45#card-21810125

File: src/Tools/ExternalAccess/LiveShare/Projects/RoslynRemoteProjectInfoProvider.cs (hunk 13 above)
Comment (dibarbet) [reply src/Tools/ExternalAccess/LiveShare/Projects/RoslynRemoteProjectInfoProvider.cs]: Yep this is not part of LSP.  Have a work item for it
https://github.com/dotnet/roslyn/projects/45#card-21810235

File: src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs (hunk 4 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs]: Did the second - this is only used in one place so I'm not sure it needs an extension method.  New handlers should use the actual client capability

File: src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs (hunk 6 above)
Comment (dibarbet) [reply src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs]: That doesn't seem any simpler, I'd first convert to a jtoken (JToken.FromObject), set the property, then convert it back to an actual client capability type to pass it in.  Seems easy enough to just create it like I did above.  

I can move it to a helper method, but this shouldn't be used anywhere but here.

File: src/Tools/ExternalAccess/LiveShare/StringConstants.cs [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare
+{
+    internal class StringConstants
+    {
+        public const string BaseRemoteAssemblyTitle = "Base Remote Language Service";
+
+        // The service name for an LSP server implemented using Roslyn designed to be used with the Roslyn client
+        public const string RoslynContractName = "Roslyn";
+        // The service name for an LSP server implemented using Roslyn designed to be used with the LSP SDK client
+        public const string RoslynLSPSDKContractName = "RoslynLSPSDK";
```
Comment (tmat) [inline src/Tools/ExternalAccess/LiveShare/StringConstants.cs]: >RoslynLSPSDKContractName [](start = 28, length = 24)

Style: Naming guidelines for 3+ letter abbreviations are to capitalize first letter and make the rest lower case. So, this should be `RoslynLspSdkContractName`

File: src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.cs [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServices.LiveShare.CustomProtocol;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Classification
+{
+    internal class RoslynClassificationService : ISyntaxClassificationService
+    {
+        private readonly RoslynLSPClientServiceFactory _roslynLSPClientServiceFactory;
```
Comment (tmat) [inline src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.cs]: >RoslynLSPClientServiceFactory [](start = 25, length = 29)

Style: `RoslynLspClientServiceFactory`

File: src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+    [ExportLanguageServiceFactory(typeof(ISyntaxClassificationService), StringConstants.VBLspLanguageName), Shared]
+    internal class VBLspEditorClassificationFactoryService : ILanguageServiceFactory
+    {
+        private readonly RoslynLSPClientServiceFactory _roslynLSPClientServiceFactory;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        public VBLspEditorClassificationFactoryService(RoslynLSPClientServiceFactory roslynLSPClientServiceFactory, ClassificationTypeMap classificationTypeMap, IThreadingContext threadingContext)
```
Comment (tmat) [inline src/Tools/ExternalAccess/LiveShare/Classification/RoslynClassificationService.Exports.cs]: >roslynLSPClientServiceFactory [](start = 85, length = 29)

`roslynLspClientServiceFactory`

**Comment by dibarbet** [comment]
@tmat any further comments I should address before merging?

**Comment by tmat** [comment]
@dibarbet If @heejaechang signs off then I'm fine with the change.

**Pull request merged**